lightningd/lightningd.c

```diff
@@ -573,11 +575,18 @@ void notify_new_block(struct lightningd *ld, u32 block_height)
 	channel_notify_new_block(ld, block_height);
 }
 
+/*~ Signal handler. Shuts down by breaking out of io loop. */
+static void shutdown_lightningd(int signum)
+{
+	log_info(ld->log, "Received signal %d, shutting down.", signum);
```

Comment:
Is `log_info` reentrant? It is theoretically possible that `log_info` (or a function manipulating same data structures) is executing when a signal is delivered.  Calling it again while it is already executing may leave its data structures in an invalid state, leading to problems as we try to clean it up while exiting.

I suggest using a Posix `pipe` as I mentioned.  This is known as the self-pipe trick:

* https://www.sitepoint.com/the-self-pipe-trick-explained/
* https://lwn.net/Articles/177897/
* http://man7.org/tlpi/code/online/diff/altio/self_pipe.c.html

Reply:
I appreciate the feedback.  I'll get to work on the pipe approach.